### DISCUSSION TITLE
Grab-bag of Admin UI Customisation Requests

### DISCUSSION THREAD

**Discussion by JedWatson** [keystonejs/keystone#5309] [Feature requests]
In #4260 @gautamsi shared some requests and use-cases he has for things in the Admin UI, I wanted to start a thread to collect more of them that we can take into account while working on the new Admin UI design and customisation plan.

So if you have requirements like this (or have done / needed things in projects) please thread them here!

(see the linked PR for context on the UI change being discussed)

---

Some thought on this based on my past experience with UI Hooks

* Placement like this will affect number of primary actions which can be placed on Item List page, I can visualize that it will be ugly when we want to add more item there. 
* if you place this filter and primary action together, Where would you add search. 
    * one option would be to add a global search on header section which can search item based on context (when search is fixed) then it can have GitHub like search context like this
![image](https://user-images.githubusercontent.com/5769869/99193893-d6174800-27a1-11eb-91ad-c839394ba808.png)

* I do have a customer where I have added 4-5 primary action and some secondary actions (when you select items) to be applied on selected items

* I like the idea to add "Save" button on top of Create Item modal, I noticed that we have some view hints in ui config for items, one of the way we can also minimize the number of fields needed for initial create item view, the sequence is also important when rendered.

* I also have request from customer to have full page editing like item details page rather than slide out modal. Basically I have request for both the form being either full page or both like create item modal.

it would be helpful if we can preserve the extensibility to customize the form rather than compose ourself. I know the focus is on composability rather than the ui hooks in new interface.

**Comment by VinayaSathyanarayana** [comment]
Need info on how to add a custom Page to Admin Screen
One should be able to configure Access Control for the Custom Page

**Reply by raveling** [reply]
@VinayaSathyanarayana

FYI, we have some guidance on adding custom pages to Admin UI now:

https://keystonejs.com/docs/guides/custom-admin-ui-pages
https://keystonejs.com/docs/guides/custom-admin-ui-navigation

**Comment by VinayaSathyanarayana** [comment]
- [ ]  Custom Actions keystonejs/keystone#4888 

- [ ] Storybook: https://storybook.js.org/ 

- [ ] RBAC - Role Based Access Control - leveraging the  nice List and Field Level access control already available in Keystone. - Need a way to add roles

- [ ] Customizing the format of the List View
* Field formatters and editors - [Built-In,  Custom, Community Add-ons] . 
References:
 As explained in https://www.keystonejs.com/blog/roadmap-update
A number of good ideas are available from - http://tabulator.info/docs/4.0/format
* Row Height
* Column Width
* Tool Tips
* Position of Standard Buttons + Custom Actions

**Comment by VinayaSathyanarayana** [comment]
- [ ] Support / Integrate Prisma Studio

**Comment by VinayaSathyanarayana** [comment]
- [ ]  Workflow / State Machine / Conditional Options for the `Select` Field

-  The List of options should come from a function at run time and not be decided at development time.
- This will be essential for supporting workflows using state machines like [xstate](https://github.com/davidkpiano/xstate)
- This will also be useful for simple cases like first you select country and then you select state and then you select city

``` javascript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { select } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: select({
          dataType: 'enum',
          options: (context) => {
             /* logic to check current state and decide on valid set of options */
             return ([
              { label: '...', value: '...' },
              /* ... */
            ]);
          },
          defaultValue: '...',
          isRequired: true,
          isIndexed: true,
          isUnique: true,
          ui: { displayMode: 'select' },
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

@JedWatson - Request you look into this.

**Comment by VinayaSathyanarayana** [comment]
Build time `hooks` /  `extensions` / `plug-ins` to cover use cases like

- [ ] Generate `OpenAPI` Specs from Keystone List configuration
- [ ] Generate `PlantUML` Diagrams from prima schema
- [ ] Generate `sitemap`
- [ ] .... Others

**Comment by VinayaSathyanarayana** [comment]
- [ ] Support [Open Tracing](https://opentracing.io/docs/best-practices/instrumenting-frameworks/)

**Comment by VinayaSathyanarayana** [comment]
Any updates?

**Reply by VinayaSathyanarayana** [reply]
Would appreciate some inputs on this list
